File: fast-linear-estimator/src/exp_approx.rs
const EXP_BIAS_32: i32 = 127; // zero point for exponent

pub mod exp_f32_const {
    // taken from cephes/avxfun
    pub const EXP_HI: f32 = 88.3762626647949;
    pub const EXP_LO: f32 = -88.3762626647949;
    // reduced negative limit because we're converting to signed integers
    pub const EXP_LO_AVX_SIGNED: f32 = -88.028;

    // taken from inavec
    pub const C0: f32 = 1.06906116358144185133e-04;
    pub const C1: f32 = 3.03543677780836240743e-01;
    pub const C2: f32 = -2.24339532327269441936e-01;
    pub const C3: f32 = -7.92041454535668681958e-02;

    pub const S: f32 = (1u64 << 23) as f32;
    pub const B: f32 = S * (super::EXP_BIAS_32 as f32);
}

#[allow(dead_code)]
pub fn exp_approx_f32(x_in: f32) -> f32 {
    // clamp x
    let mut x = x_in;
    x = x.min(exp_f32_const::EXP_HI);
    x = x.max(exp_f32_const::EXP_LO);

    // apply approximation
    x = x * std::f32::consts::LOG2_E;
    let fl = x.floor();
    let xf = x - fl;

    let mut kn = exp_f32_const::C3;
    kn = xf * kn + exp_f32_const::C2;
    kn = xf * kn + exp_f32_const::C1;
    kn = xf * kn + exp_f32_const::C0;
    x -= kn;

    // create integer with bits in the right place, by rounding double to integer,
    // then re-interpret as a double
    let xul = ((exp_f32_const::S * x) + exp_f32_const::B) as u32;
    let res = unsafe { std::mem::transmute::<u32, f32>(xul) };
    res
}

#[cfg(test)]
mod tests {

    use approx::*;

    const VALS: [f32; 8] = [-10_f32, -5., -1., 0., 1., 2., 5., 10.];

    fn expected() -> Vec<f32> {
        VALS.iter().map(|v| v.exp()).collect()
    }

    fn check_assert(expect: &[f32], vals: &[f32]) {
        vals.iter().zip(expect.iter()).for_each(|(act, exp)| {
            assert_relative_eq!(exp, act, max_relative = 1e-4);
        });
    }

    fn check_assert_accurate(expect: &[f32], vals: &[f32]) {
        vals.iter().zip(expect.iter()).for_each(|(act, exp)| {
            assert_relative_eq!(exp, act, max_relative = 1e-8);
        });
    }

    #[test]
    fn exp_approx_f32() {
        let res: Vec<_> = VALS.iter().map(|&v| super::exp_approx_f32(v)).collect();
        let expect = expected();
        check_assert(&expect, &res);
    }

    #[cfg(target_arch = "x86_64")]
    #[test]
    fn exp_approx_avxf32() {
        use std::arch::x86_64::*;

        let input: __m256 = unsafe { _mm256_loadu_ps(&VALS[0]) };
        let res = crate::exp_approx_avx::exp_approx_avxf32(input);

        let res_f32: [f32; 8] = unsafe { std::mem::transmute(res) };
        check_assert(&expected(), &res_f32);
    }

    #[cfg(target_arch = "x86_64")]
    #[test]
    fn exp_sleef_avx() {
        use std::arch::x86_64::*;

        let input: __m256 = unsafe { _mm256_loadu_ps(&VALS[0]) };
        let res = crate::exp_sleef_avx::exp(input);

        let res_f32: [f32; 8] = unsafe { std::mem::transmute(res) };
        check_assert_accurate(&expected(), &res_f32);
    }

    #[cfg(target_arch = "aarch64")]
    #[test]
    fn exp_approx_armf32() {
        use std::arch::aarch64::*;

        unsafe {
            let x1: float32x4_t = std::mem::transmute([-10_f32, -5., -1., 0.]);
            let x2: float32x4_t = std::mem::transmute([1.0_f32, 2., 5., 10.]);

            let res1: [f32; 4] = std::mem::transmute(crate::exp_approx_arm::exp_approx_armf32(x1));
            let res2: [f32; 4] = std::mem::transmute(crate::exp_approx_arm::exp_approx_armf32(x2));

            let expect = expected();
            check_assert(&expect[0..4], &res1);
            check_assert(&expect[4..8], &res2);
        }
    }
}

File: csharp/FastLinearEstimator/ExpApprox.cs
using System;
using System.Diagnostics;
using System.Numerics;
using System.Runtime.CompilerServices;
using System.Runtime.Intrinsics;
using System.Runtime.Intrinsics.X86;

namespace FastLinearEstimator
{
    public static class ExpApprox
    {
        public const float LOG2_E = 1.44269504088896340735992468100189214f;
        public const int EXP_BIAS_32 = 127;
        public const float EXP_HI = 88.3762626647949f;
        public const float EXP_LO = -88.3762626647949f;
        // reduced negative limit because we're converting to signed integers
        public const float EXP_LO_AVX_SIGNED = -88.028f;
        // taken from inavec
        public const float C0 = 1.06906116358144185133e-04f;
        public const float C1 = 3.03543677780836240743e-01f;
        public const float C2 = -2.24339532327269441936e-01f;
        public const float C3 = -7.92041454535668681958e-02f;
        public const float S = (float)(1u << 23);
        public const float B = S * (float)EXP_BIAS_32;


        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public static Vector256<float> ExpApproxAvx(Vector256<float> vector)
        {
            var x = vector;
            x = Avx.Min(x, Vector256.Create(EXP_HI));
            x = Avx.Max(x, Vector256.Create(EXP_LO_AVX_SIGNED));

            x = Avx.Multiply(x, Vector256.Create(LOG2_E));
            var fl = Avx.Floor(x);
            var xf = Avx.Subtract(x, fl);

            var kn = Vector256.Create(C3);
            kn = Avx.Add(Avx.Multiply(xf, kn), Vector256.Create(C2));
            kn = Avx.Add(Avx.Multiply(xf, kn), Vector256.Create(C1));
            kn = Avx.Add(Avx.Multiply(xf, kn), Vector256.Create(C0));
            x = Avx.Subtract(x, kn);

            var xf32 = Avx.Add(
                Avx.Multiply(Vector256.Create(S), x),
                Vector256.Create(B)
            );
            var xul = Avx.ConvertToVector256Int32(xf32);
            var res = xul.AsSingle();
            return res;
        }

        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public static Vector<float> ExpApproxVector(Vector<float> vector)
        {
            var x = vector;
            x = Vector.Min(x, new Vector<float>(EXP_HI));
            x = Vector.Max(x, new Vector<float>(EXP_LO_AVX_SIGNED));

            x = Vector.Multiply(x, new Vector<float>(LOG2_E));

            var fl = Vector.Floor(x);
            var xf = Vector.Subtract(x, fl);

            var kn = new Vector<float>(C3);
            kn = Vector.Add(Vector.Multiply(xf, kn), new Vector<float>(C2));
            kn = Vector.Add(Vector.Multiply(xf, kn), new Vector<float>(C1));
            kn = Vector.Add(Vector.Multiply(xf, kn), new Vector<float>(C0));
            x = Vector.Subtract(x, kn);

            var xf32 = Vector.Add(
                Vector.Multiply(new Vector<float>(S), x),
                new Vector<float>(B)
            );
            var xul = Vector.ConvertToInt32(xf32);
            var res = Vector.AsVectorSingle(xul);
            return res;
        }
    }
}